### PR TITLE
APP-1389: add bitbucket data center resolver for pr comment triggers

### LINKED ISSUE #14095: Implement Bitbucket Data Center webhook handler for @OpenHands resolver
## Summary

The `@OpenHands` resolver does not trigger when commenting on Bitbucket Data Center pull requests. While PR #11052 added Bitbucket Data Center authentication and API services, the webhook handler infrastructure needed to receive and process PR comment events was not implemented.

### What Works ✅
- Bitbucket DC OAuth login via Keycloak
- Bitbucket DC API service (`SaaSBitbucketDCService`)
- Bitbucket DC proxy for user info (`/bitbucket-dc-proxy/oauth2/userinfo`)
- Bitbucket DC resolver mixin (helper methods in `openhands.integrations.bitbucket_data_center.service.resolver`)

### What Is Missing ❌
- **Webhook endpoint**: No `/integration/bitbucket-dc/events` route exists
- **Webhook router**: No `server/routes/integration/bitbucket_dc.py` file
- **Webhook manager**: No `integrations/bitbucket_data_center/bitbucket_dc_manager.py` class
- **Ingress configuration**: No ingress route for Bitbucket DC webhook endpoint

## Investigation Details

Looking at the current ingress configuration (`openhands-integrations-ingress`):

```yaml
paths:
  - path: /integration/github/events    # ✅ exists
  - path: /integration/gitlab/events    # ✅ exists
  - path: /integration/jira/events      # ✅ exists
  # /integration/bitbucket-dc/events    # ❌ MISSING
```

Integration routers in `/app/server/routes/integration/`:
- `github.py` ✅
- `gitlab.py` ✅
- `jira.py` / `jira_dc.py` ✅
- `linear.py` ✅
- `slack.py` ✅
- `bitbucket_dc.py` ❌ **MISSING**

### 1. Create Webhook Router
`server/routes/integration/bitbucket_dc.py`:
- Endpoint: `POST /integration/bitbucket-dc/events`
- Verify webhook signature/secret
- Parse Bitbucket DC webhook payload format
- Route events to the manager

### 2. Create Webhook Manager
`integrations/bitbucket_data_center/bitbucket_dc_manager.py`:
- Process incoming webhook events
- Handle PR comment events (specifically `@OpenHands` mentions)
- Trigger resolver conversations
- Similar pattern to `github_manager.py` and `gitlab_manager.py`

### 3. Update Server Configuration
`saas_server.py`:
- Add conditional router registration similar to other integrations:

```python
if BITBUCKET_DATA_CENTER_HOST:
    from server.routes.integration.bitbucket_dc import bitbucket_dc_integration_router
    base_app.include_router(bitbucket_dc_integration_router)
```

### 4. Update Helm Chart
Add webhook endpoint to `openhands-integrations-ingress`:

```yaml
- backend:
    service:
      name: openhands-integrations-service
      port:
        number: 3000
  path: /integration/bitbucket-dc/events
  pathType: Exact
```

### 5. Environment Variables
May need new env vars:
- `BITBUCKET_DATA_CENTER_WEBHOOK_SECRET` - for webhook signature verification
- `BITBUCKET_DC_WEBHOOKS_ENABLED` - feature flag (optional)

## Bitbucket DC Webhook Event Format

Bitbucket Data Center sends webhook events with different payload structure than GitHub/GitLab. Key events to handle:
- `pr:comment:added` - When a comment is added to a PR
- `pr:comment:edited` - When a comment is edited

Reference: [Bitbucket Data Center Webhooks Documentation](https://confluence.atlassian.com/bitbucketserver/event-payload-938025882.html)

## Testing Checklist
- [ ] Webhook endpoint receives events from Bitbucket DC
- [ ] Signature verification works correctly
- [ ] PR comment events with `@OpenHands` trigger resolver
- [ ] Resolver creates conversation and responds on PR
- [ ] Error handling for malformed payloads

## Related
- PR that added BDC support: #11052
- Similar implementation reference: `server/routes/integration/github.py`

### PR DESCRIPTION
**Pull request by hieptl**
- [x] A human has tested these changes.

---

## Why

Self-hosted customers can't use the Cloud resolver — payload shape, REST base, auth model, and permission API all differ; the most load-bearing ones:

| Concern | Cloud | DC |
|---|---|---|
| API base | `api.bitbucket.org/2.0` | `{host}/rest/api/1.0` |
| PR-comment event | `pullrequest:comment_created` | `pr:comment:added` (+ edited) |
| Comment body field | `comment.content.raw` | `comment.text` |
| Repo identity in payload | `repository.full_name` | `pullRequest.toRef.repository.{slug, project.key}` |
| Inline anchor | `comment.inline.{path,to/from}` | `comment.anchor.{path,line,lineType,fileType}` |
| Reply payload | `{content:{raw}, parent, inline}` | `{text, parent, anchor}` |
| Token refresh | OAuth refresh supported | DC: no refresh — long-lived OAuth access tokens |
| Timestamps | ISO 8601 | epoch milliseconds |

## Summary

Adds a Bitbucket Data Center PR-comment trigger that mirrors the Bitbucket Cloud resolver (#14205) for self-hosted BBDC instances. A `@openhands` mention in a PR comment dispatches a job; the agent's reply lands as a threaded (or anchored, for inline comments) reply on the same PR.

DC is isolated from Cloud throughout: its own `SourceType` enum value, manager, view hierarchy, webhook route, storage table, and migration — same separation already used for `jira` vs `jira_dc`.

#### App-server (OSS)

- `openhands/app_server/integrations/bitbucket_data_center/service/resolver.py`
  - New: `reply_to_pr_comment(owner, repo_slug, pr_id, body, *, parent_comment_id, anchor)`. Posts to `…/pull-requests/{id}/comments` with `text`, optional `parent`, optional `anchor` for inline replies.
  - New: `user_has_write_access`, `user_has_write_access_for`. Policy explained in [Permission gating](#permission-gating) below.
- `openhands/app_server/integrations/bitbucket_data_center/bitbucket_dc_service.py`
  - `BASE_URL` falls back to the `BITBUCKET_DATA_CENTER_HOST` env var when `base_domain` is not passed. Without this, every SaaS-side instantiation (`SaaSBitbucketDCService(external_auth_id=…)`) ended up with `BASE_URL=''` and httpx rejected schemeless URLs with `UnsupportedProtocol`.
- `openhands/app_server/integrations/bitbucket_data_center/service/base.py`
  - `_get_headers` lazy-loads `self.token` via `get_latest_token()` when empty (the default for SaaS construction with `external_auth_id`), mirroring the Cloud base. Without this the header was `Basic ` with an empty base64 payload — httpx rejected it as `Illegal header value`.
  - Chooses **Basic** vs **Bearer** by token format: colon-containing tokens (`x-token-auth:<PAT>`, `user:pass`) → Basic; raw OAuth 2.0 access tokens → Bearer per RFC 6750.

### Enterprise

- `enterprise/integrations/models.py` — adds `SourceType.BITBUCKET_DATA_CENTER`.
- `enterprise/integrations/bitbucket_data_center/`
  - `bitbucket_dc_view.py` — `BitbucketDCPR`, `BitbucketDCPRComment`, `BitbucketDCInlinePRComment`, and `BitbucketDCFactory` with `is_pr_comment` and `create_bitbucket_dc_view_from_payload`. `PR_COMMENT_EVENTS = {'pr:comment:added', 'pr:comment:edited'}`.
  - `bitbucket_dc_manager.py` — `BitbucketDCManager` mirroring the Cloud manager. Resolves the IdP token via `ProviderType.BITBUCKET_DATA_CENTER`, looks up the installer by `(project_key, repo_slug)` from the new webhook store.
  - `bitbucket_dc_v1_callback_processor.py` — posts the agent's summary comment back via `SaaSBitbucketDCService.reply_to_pr_comment`.
- `enterprise/server/routes/integration/bitbucket_dc.py` — `POST /integration/bitbucket-dc/events`. HMAC-SHA256 verification against the per-repo secret (looked up from `bitbucket_dc_webhook`). Repo identity read from `payload.pullRequest.toRef.repository.{project.key, slug}`. Honors `diagnostics:ping` (DC's webhook test event) with a 200 short-circuit. Dedup key `bbdc:{event_key}:{pr_id}:{comment_id}:{request_id}` with a body-hash fallback.
- `enterprise/storage/bitbucket_dc_webhook.{py,_store.py}` — new SQLAlchemy model and store. Unique constraint on `(project_key, repo_slug)`.
- `enterprise/migrations/versions/112_create_bitbucket_dc_webhook_table.py` — alembic revision `112`, `down_revision='111'`.
- `enterprise/saas_server.py` — registers the resolver router and DC proxy under the existing `BITBUCKET_DATA_CENTER_HOST` env-var gate.

### Templates

Reuses the Cloud Jinja templates at `openhands/app_server/integrations/templates/resolver/bitbucket/`. The template variables are provider-agnostic, so duplicating them for DC would be churn. The DC view's `_get_instructions` loads from the same folder.

### Out of scope (follow-ups)

- UI flow for installing DC webhooks (today the row must be seeded by SQL and the webhook configured by hand in BBDC admin).
- Allowlist-based commenter gating beyond "actor identifiable in payload" (BBDC's OAuth grant doesn't expose permission introspection without admin scope; if customers want stricter gating, that lives on top).
- Hardening the OpenHands-Cloud chart to fail loudly when `bitbucketDataCenter.enabled=true` but `host` is missing (currently silently stamps `placeholder.invalid` into Keycloak).

## How to Test

1. Please run the SaaS Code locally.
2. Sign in using your Bitbucket Data Center account.
3. Navigate to a Bitbucket repository and add a comment to trigger the resolver.

## Video/Screenshots

https://github.com/user-attachments/assets/8e625bef-773d-49cd-a7fd-72132c3a109e

## Type

- [ ] Bug fix
- [x] Feature
- [ ] Refactor
- [ ] Breaking change
- [ ] Docs / chore

---

To run this PR locally, use the following command:

GUI with Docker:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   --name openhands-app-88e81ea   docker.openhands.dev/openhands/openhands:88e81ea
```

Resolves https://github.com/OpenHands/OpenHands/issues/14095

### CODE REVIEW

File: openhands/app_server/integrations/bitbucket_data_center/service/resolver.py [hunk 1]
```diff
@@ -111,3 +111,77 @@ def _process_raw_comments(
...
+        customer's BBDC admin to grant the integration a permission it
+        never actually needs to do its work.
+
+        Mirroring the Cloud counterpart's "downgrade to a non-admin
+        endpoint" pattern (see
+        ``openhands/app_server/integrations/bitbucket/service/resolver.py``),
+        we trust the act of webhook installation as the implicit
+        endorsement of anyone who can interact with PRs in the repo. The
+        trigger is gated by:
+
```
[the remark this answers is not in the document]
Comment (hieptl) [reply openhands/app_server/integrations/bitbucket_data_center/service/resolver.py:176]: **Decision:** **Not addressed in this PR** (intentional, recorded as out-of-scope)

**Rationale**

The relaxed permission policy is a deliberate design call discussed in the PR description's "Permission gating" section. Its reasoning is captured in the docstring at lines 154-186:

- Bitbucket DC's OAuth 2.0 provider has **no** permission-introspection endpoint that works without `REPO_ADMIN` scope (every variant of `/permissions/users` requires admin per Atlassian's docs).
- The OpenHands grant requests `REPO_WRITE`, which is sufficient for everything the agent does (read PRs, post comments).
- Forcing customers to bump their Application Link to `REPO_ADMIN` just so we can call an endpoint we don't otherwise need is a worse trade-off than relaxing the gate.
- The two gates that remain — **HMAC-signed webhook delivery** and **identifiable actor in payload** — match the Cloud counterpart's downgrade pattern at `openhands/app_server/integrations/bitbucket/service/resolver.py:151-158`.

A `BITBUCKET_DC_ALLOWED_USERS` env var is a reasonable follow-up feature, but it's its own design call (env var vs DB-backed allowlist; per-repo vs per-installation; group support; etc.) that warrants a separate PR with security review. The PR description and docstring both flag this as an explicit follow-up:

> *"Customers needing finer-grained gating should layer a separate allowlist on top — the resolver doesn't ship one today."*

Adding an env-var allowlist here would be scope creep on a security-sensitive design. Tracking as a follow-up ticket is the right move.

Thank you! 🙏

File: openhands/app_server/integrations/bitbucket_data_center/bitbucket_dc_service.py [hunk 2]
```diff
@@ -44,6 +44,12 @@ def __init__(
         external_token_manager: bool = False,
         base_domain: str | None = None,
     ) -> None:
+        # Fall back to the BITBUCKET_DATA_CENTER_HOST env var when no domain
+        # is passed explicitly — call sites in the SaaS resolver path
+        # construct this service without base_domain, and an empty BASE_URL
+        # silently produces schemeless API URLs that httpx rejects.
+        if not base_domain:
```
[the remark this answers is not in the document]
Comment (hieptl) [reply openhands/app_server/integrations/bitbucket_data_center/bitbucket_dc_service.py:51]: **Decision:** **Not addressed**

**Rationale**

The env-var fallback is the *canonical config knob* for this service, not a workaround. Three reasons:

1. **The Cloud counterpart has the same shape**: `openhands/app_server/integrations/bitbucket/bitbucket_service.py:51` uses `self.base_domain = base_domain or 'bitbucket.org'` — Cloud has a hardcoded global default; DC has no such default, so the env var fills that role. `BITBUCKET_DATA_CENTER_HOST` is already the canonical host config: it gates the SaaS proxy router (`enterprise/saas_server.py:154`), is read by `enterprise/server/auth/constants.py:51`, and is wired through the chart's `_env.yaml`. Reading it once at construction matches that contract.

2. **Making `base_domain` required spreads the env-var read across call sites**, not removes it. There are 3+ SaaS construction points (`bitbucket_dc_manager.py:84` and `:144`, `bitbucket_dc_v1_callback_processor.py:96`, plus future ones). Each would have to do `os.environ.get('BITBUCKET_DATA_CENTER_HOST')` itself — that's *more* coupling, not less.

3. **It's not a breaking change**: `base_domain` stays optional, callers that pass it explicitly still win, and the env-var path keeps the OSS class usable for any consumer (including tests that intentionally omit it — see `test_init_no_domain` and `test_init_falls_back_to_env_var_when_base_domain_missing` in `tests/unit/integrations/bitbucket_data_center/test_bitbucket_dc.py:23-39`).

The fallback's behavior is also asserted by the regression tests that landed in this PR — explicit `base_domain` overrides the env, missing both produces an empty `BASE_URL` (preserving the historical signal). The fallback is documented and tested, not implicit.

Thank you! 🙏

File: openhands/app_server/integrations/bitbucket_data_center/service/base.py [hunk 3]
```diff
@@ -67,12 +67,34 @@ def _has_token_expired(self, status_code: int) -> bool:
...
+        payload and httpx rejects it as ``Illegal header value``.
+
+        Tokens in ``username:secret`` form (e.g. ``x-token-auth:<PAT>`` or
+        ``user:pass``) use HTTP Basic auth. Raw OAuth 2.0 access tokens
+        (no colon) use Bearer per RFC 6750 — this is the format the SaaS
+        broker flow returns from Keycloak.
+        """
+        if not self.token or not self.token.get_secret_value():
+            latest_token = await self.get_latest_token()
+            if latest_token:
```
[the remark this answers is not in the document]
Comment (hieptl) [reply openhands/app_server/integrations/bitbucket_data_center/service/base.py:85]: **Decision:** **Not addressed in this PR** (deliberate symmetry with Cloud)

**Rationale**

The DC base intentionally mirrors the Bitbucket Cloud base byte-for-byte in this area. Compare:

- DC: `openhands/app_server/integrations/bitbucket_data_center/service/base.py:69-99`
- Cloud: `openhands/app_server/integrations/bitbucket/service/base.py:64-90`

Both inline the lazy-load + header-format logic in `_get_headers`. We mirror this exactly so the two services remain symmetric for future maintenance — when one is refactored (e.g., if Cloud later adopts the `_ensure_token_loaded()` extraction), the other should follow in the same change.

The suggested extraction is a clean stylistic improvement, but applying it to DC alone introduces asymmetry that costs more to maintain than it saves. If reviewers feel strongly, the right place to land it is a follow-up PR that updates **both** Cloud and DC bases at once. I'd be happy to open that as a separate housekeeping PR.

Thank you! 🙏

File: enterprise/server/routes/integration/bitbucket_dc.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            project_key=project_key,
+            repo_slug=repo_slug,
+        )
+
+        pr_id = (payload_data.get('pullRequest') or {}).get('id')
+        comment_id = (payload_data.get('comment') or {}).get('id')
+
+        if x_request_id:
+            dedup_key = f'bbdc:{x_event_key}:{pr_id}:{comment_id}:{x_request_id}'
+        else:
```
[the remark this answers is not in the document]
Comment (hieptl) [reply enterprise/server/routes/integration/bitbucket_dc.py:125]: I have updated the code based on your feedback. Thank you! 🙏

File: enterprise/server/routes/integration/bitbucket_dc.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            status_code=200,
+            content={
+                'message': 'Bitbucket DC events endpoint reached successfully.',
+            },
+        )
+
+    except HTTPException:
+        raise
+    except Exception as e:
+        logger.exception(f'Error processing Bitbucket DC event: {e}')
```
[the remark this answers is not in the document]
Comment (hieptl) [reply enterprise/server/routes/integration/bitbucket_dc.py:158]: I have updated the code based on your feedback. Thank you! 🙏

File: enterprise/integrations/bitbucket_data_center/bitbucket_dc_manager.py [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+        self,
+        token_manager: TokenManager,
+        data_collector: None = None,
+    ) -> None:
+        self.token_manager = token_manager
+        self.webhook_store = BitbucketDCWebhookStore()
+        # Reuse the Bitbucket Cloud Jinja templates — wording is provider-
+        # agnostic and the variables (pr_number, branch_name, comments,
+        # file_location, line_number) match.
+        self.jinja_env = Environment(
```
[the remark this answers is not in the document]
Comment (hieptl) [reply enterprise/integrations/bitbucket_data_center/bitbucket_dc_manager.py:50]: **Decision:** **Not addressed in this PR**

**Rationale**

The new DC manager matches the established pattern across **every** existing manager in this codebase:

| Manager | Path construction |
|---|---|
| `enterprise/integrations/github/github_manager.py:56` | `OPENHANDS_RESOLVER_TEMPLATES_DIR + 'github'` |
| `enterprise/integrations/gitlab/gitlab_manager.py:44` | `OPENHANDS_RESOLVER_TEMPLATES_DIR + 'gitlab'` |
| `enterprise/integrations/jira/jira_manager.py:71` | `OPENHANDS_RESOLVER_TEMPLATES_DIR + 'jira'` |
| `enterprise/integrations/jira_dc/jira_dc_manager.py:50` | `OPENHANDS_RESOLVER_TEMPLATES_DIR + 'jira_dc'` |
| `enterprise/integrations/slack/slack_manager.py:70` | `OPENHANDS_RESOLVER_TEMPLATES_DIR + 'slack'` |
| `enterprise/integrations/bitbucket/bitbucket_manager.py:47` | `OPENHANDS_RESOLVER_TEMPLATES_DIR + 'bitbucket'` |

`OPENHANDS_RESOLVER_TEMPLATES_DIR` is defined with a trailing slash (`enterprise/integrations/utils.py:66-69`), so `+ 'name'` produces the correct path on every platform the SaaS actually runs on (Linux containers, macOS dev boxes). Changing only the DC manager would create an inconsistency that's harder to reason about than the current uniform pattern.

If the codebase wants to migrate to `os.path.join` (or `pathlib.Path`), that's a sensible global housekeeping pass — but it should touch all 7 managers in a single change, not one. Out of scope for a feature PR.

Thank you! 🙏

File: enterprise/integrations/bitbucket_data_center/bitbucket_dc_view.py [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+    @staticmethod
+    async def create_bitbucket_dc_view_from_payload(
+        message: Message, keycloak_user_id: str
+    ) -> BitbucketDCViewType:
+        """Build a view from a Bitbucket DC webhook payload.
+
+        ``keycloak_user_id`` is the OpenHands user that **installed** the
+        webhook (looked up by ``(project_key, repo_slug)`` from the
+        ``bitbucket_dc_webhook`` table by the manager). The resolver acts
+        on Bitbucket DC as the installer.
```
[the remark this answers is not in the document]
Comment (hieptl) [reply enterprise/integrations/bitbucket_data_center/bitbucket_dc_view.py:290]: I have updated the code based on your feedback. Thank you! 🙏

**Review by tofarr — APPROVED** [review]
